File: src/lib/db/digests.ts
// Database operations for digests table
import { randomUUID } from 'crypto';
import { getDatabase } from './connection';
import { getLogger } from '@/lib/log/logger';
import type { Digest, DigestRecordRow } from '@/types';

const log = getLogger({ module: 'DBDigests' });

function formatStack(maxLines: number = 5): string | undefined {
  const stack = new Error().stack;
  if (!stack) return undefined;
  return stack
    .split('\n')
    .slice(2, 2 + maxLines)
    .map(line => line.trim())
    .join(' | ');
}

/**
 * Create a new digest in the database
 */
export function createDigest(digest: Digest): void {
  const db = getDatabase();

  const stmt = db.prepare(`
    INSERT INTO digests (
      id, file_path, digester, status, content, sqlar_name, error,
      attempts, created_at, updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ON CONFLICT(id) DO UPDATE SET
      file_path = excluded.file_path,
      digester = excluded.digester,
      status = excluded.status,
      content = excluded.content,
      sqlar_name = excluded.sqlar_name,
      error = excluded.error,
      attempts = excluded.attempts,
      updated_at = excluded.updated_at
  `);

  stmt.run(
    digest.id,
    digest.filePath,
    digest.digester,
    digest.status,
    digest.content,
    digest.sqlarName,
    digest.error,
    digest.attempts ?? 0,
    digest.createdAt,
    digest.updatedAt
  );

  if (digest.status === 'todo') {
    log.debug(
      { filePath: digest.filePath, digester: digest.digester, attempts: digest.attempts ?? 0, stack: formatStack() },
      'createDigest set status=todo'
    );
  }
}

/**
 * Create or reset a digest to pending status (for enqueue operations)
 * If digest exists, resets it to pending state. If not, creates it.
 */
export function upsertPendingDigest(digest: Omit<Digest, 'updatedAt'>): void {
  const db = getDatabase();

  const stmt = db.prepare(`
    INSERT INTO digests (
      id, file_path, digester, status, content, sqlar_name, error,
      attempts, created_at, updated_at
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ON CONFLICT(id) DO UPDATE SET
      status = 'todo',
      error = NULL,
      attempts = 0,
      updated_at = ?
  `);

  const now = new Date().toISOString();
  stmt.run(
    digest.id,
    digest.filePath,
    digest.digester,
    digest.status,
    digest.content,
    digest.sqlarName,
    digest.error,
    0,
    digest.createdAt,
    now,
    now  // updated_at for the UPDATE clause
    );

  log.debug(
    { filePath: digest.filePath, digester: digest.digester, attempts: 0, stack: formatStack() },
    'upsertPendingDigest set status=todo'
  );
}

/**
 * Insert a digest only if it does not exist.
 * Used by sync routines to avoid overwriting existing statuses.
 */
export function insertDigestIfMissing(digest: Digest): void {
  const db = getDatabase();

  const stmt = db.prepare(`
    INSERT OR IGNORE INTO digests (
      id, file_path, digester, status, content, sqlar_name, error,
      attempts, created_at, updated_at
    ) VALUES (
      COALESCE(?, ?), ?, ?, ?, ?, ?, ?, ?, ?, ?
    )
  `);

  const result = stmt.run(
    digest.id,
    digest.id ?? randomUUID(),
    digest.filePath,
    digest.digester,
    digest.status,
    digest.content,
    digest.sqlarName,
    digest.error,
    digest.attempts ?? 0,
    digest.createdAt,
    digest.updatedAt
  );

  if (result.changes === 0) {
    log.debug(
      { filePath: digest.filePath, digester: digest.digester },
      'insertDigestIfMissing skipped existing digest'
    );
  } else if (digest.status === 'todo') {
    log.debug(
      { filePath: digest.filePath, digester: digest.digester, attempts: digest.attempts ?? 0, stack: formatStack() },
      'insertDigestIfMissing set status=todo'
    );
  }
}

/**
 * Get a digest by ID
 */
export function getDigestById(id: string): Digest | null {
  const db = getDatabase();

  const row = db
    .prepare('SELECT * FROM digests WHERE id = ?')
    .get(id) as DigestRecordRow | undefined;

  if (!row) return null;

  return rowToDigest(row);
}

/**
 * Get a digest by file path and digester name
 */
export function getDigestByPathAndDigester(filePath: string, digester: string): Digest | null {
  const db = getDatabase();

  const row = db
    .prepare('SELECT * FROM digests WHERE file_path = ? AND digester = ?')
    .get(filePath, digester) as DigestRecordRow | undefined;

  if (!row) return null;

  return rowToDigest(row);
}

/**
 * List all digests for a file
 */
export function listDigestsForPath(
  filePath: string,
  options?: {
    digesters?: string[];  // Only include these digesters
    excludeDigesters?: string[];  // Exclude these digesters
    statuses?: Digest['status'][];
    excludeStatuses?: Digest['status'][];
    order?: 'asc' | 'desc';
  }
): Digest[] {
  const db = getDatabase();

  let sql = 'SELECT * FROM digests WHERE file_path = ?';
  const params: (string | number)[] = [filePath];

  if (options?.digesters && options.digesters.length > 0) {
    const placeholders = options.digesters.map(() => '?').join(', ');
    sql += ` AND digester IN (${placeholders})`;
    params.push(...options.digesters);
  }

  if (options?.excludeDigesters && options.excludeDigesters.length > 0) {
    const placeholders = options.excludeDigesters.map(() => '?').join(', ');
    sql += ` AND digester NOT IN (${placeholders})`;
    params.push(...options.excludeDigesters);
  }

  if (options?.statuses && options.statuses.length > 0) {
    const placeholders = options.statuses.map(() => '?').join(', ');
    sql += ` AND status IN (${placeholders})`;
    params.push(...options.statuses);
  }

  if (options?.excludeStatuses && options.excludeStatuses.length > 0) {
    const placeholders = options.excludeStatuses.map(() => '?').join(', ');
    sql += ` AND status NOT IN (${placeholders})`;
    params.push(...options.excludeStatuses);
  }

  const orderDirection = options?.order === 'asc' ? 'ASC' : 'DESC';
  sql += ` ORDER BY created_at ${orderDirection}, id ${orderDirection}`;

  const rows = db
    .prepare(sql)
    .all(...params) as DigestRecordRow[];

  return rows.map(rowToDigest);
}

/**
 * List digests by digester name (across all files)
 */
export function listDigestsByDigester(digester: string, limit?: number): Digest[] {
  const db = getDatabase();

  let sql = 'SELECT * FROM digests WHERE digester = ? ORDER BY created_at DESC';
  const params: (string | number)[] = [digester];

  if (limit) {
    sql += ' LIMIT ?';
    params.push(limit);
  }

  const rows = db.prepare(sql).all(...params) as DigestRecordRow[];

  return rows.map(rowToDigest);
}

/**
 * Update a digest
 */
export function updateDigest(
  id: string,
  updates: Partial<Omit<Digest, 'id' | 'filePath' | 'createdAt'>>
): void {
  const db = getDatabase();

  const fields: string[] = [];
  const values: (string | number | null)[] = [];

  if (updates.digester !== undefined) {
    fields.push('digester = ?');
    values.push(updates.digester);
  }

  if (updates.status !== undefined) {
    fields.push('status = ?');
    values.push(updates.status);
  }

  if (updates.content !== undefined) {
    fields.push('content = ?');
    values.push(updates.content);
  }

  if (updates.sqlarName !== undefined) {
    fields.push('sqlar_name = ?');
    values.push(updates.sqlarName);
  }

  if (updates.error !== undefined) {
    fields.push('error = ?');
    values.push(updates.error);
  }

  if (updates.attempts !== undefined) {
    fields.push('attempts = ?');
    values.push(updates.attempts);
  }

  // Always update updated_at
  fields.push('updated_at = ?');
  values.push(updates.updatedAt ?? new Date().toISOString());

  values.push(id);

  const sql = `UPDATE digests SET ${fields.join(', ')} WHERE id = ?`;
  db.prepare(sql).run(...values);
}

/**
 * Delete a digest by ID
 */
export function deleteDigest(id: string): void {
  const db = getDatabase();
  db.prepare('DELETE FROM digests WHERE id = ?').run(id);
}

/**
 * Delete a specific digest by file path and digester
 */
export function deleteDigestByPathAndDigester(filePath: string, digester: string): void {
  const db = getDatabase();
  const result = db.prepare('DELETE FROM digests WHERE file_path = ? AND digester = ?').run(filePath, digester);
  if (result.changes > 0) {
    log.warn(
      { filePath, digester, removed: result.changes, stack: formatStack() },
      'deleted digest by path+digester'
    );
  }
}

/**
 * Delete all digests for a file path
 */
export function deleteDigestsForPath(filePath: string): number {
  const db = getDatabase();
  const result = db.prepare('DELETE FROM digests WHERE file_path = ?').run(filePath);
  if (result.changes > 0) {
    log.warn(
      { filePath, removed: result.changes, stack: formatStack() },
      'deleted digests for path'
    );
  }
  return result.changes;
}

/**
 * Delete all digests matching path prefix (for folder deletions)
 */
export function deleteDigestsByPrefix(pathPrefix: string): number {
  const db = getDatabase();
  const result = db.prepare('DELETE FROM digests WHERE file_path LIKE ?').run(`${pathPrefix}%`);
  if (result.changes > 0) {
    log.warn(
      { pathPrefix, removed: result.changes, stack: formatStack() },
      'deleted digests by prefix'
    );
  }
  return result.changes;
}

/**
 * Update all digest file paths (for folder renames)
 */
export function updateDigestPaths(oldPath: string, newPath: string): void {
  const db = getDatabase();
  db.prepare('UPDATE digests SET file_path = ? WHERE file_path = ?').run(newPath, oldPath);
}

/**
 * Check if a digest exists for a file path and digester
 */
export function digestExists(filePath: string, digester: string): boolean {
  const db = getDatabase();

  const row = db
    .prepare('SELECT 1 FROM digests WHERE file_path = ? AND digester = ? LIMIT 1')
    .get(filePath, digester);

  return row !== undefined;
}

/**
 * Start a digest (create with todo status)
 * This is the preferred way to initiate digest processing
 */
export function startDigest(filePath: string, digester: string): Digest {
  const existing = getDigestByPathAndDigester(filePath, digester);
  const now = new Date().toISOString();

  const digest: Digest = {
    id: existing?.id ?? randomUUID(),
    filePath,
    digester,
    status: 'todo',
    content: null,
    sqlarName: null,
    error: null,
    attempts: 0,
    createdAt: now,
    updatedAt: now,
  };

  upsertPendingDigest(digest);
  return digest;
}

/**
 * Generate a digest ID from file path and digester name
 * Format: {hash(filePath)}-{digester}
 */
export function generateDigestId(filePath: string, digester: string): string {
  const existing = getDigestByPathAndDigester(filePath, digester);
  return existing?.id ?? randomUUID();
}

/**
 * Convert database record to Digest
 */
function rowToDigest(record: DigestRecordRow): Digest {
  return {
    id: record.id,
    filePath: record.file_path,
    digester: record.digester,
    status: record.status as Digest['status'],
    content: record.content,
    sqlarName: record.sqlar_name,
    error: record.error,
    attempts: record.attempts ?? 0,
    createdAt: record.created_at,
    updatedAt: record.updated_at,
  };
}

File: backend/db/digests.go
package db

import (
	"context"
	"database/sql"

	"github.com/google/uuid"
)

// GetDigestByID retrieves a digest by ID
func (d *DB) GetDigestByID(id string) (*Digest, error) {
	query := `
		SELECT id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at
		FROM digests
		WHERE id = ?
	`

	row := d.conn.QueryRow(query, id)

	var dg Digest
	var content, sqlarName, digestError sql.NullString

	err := row.Scan(
		&dg.ID, &dg.FilePath, &dg.Digester, &dg.Status, &content,
		&sqlarName, &digestError, &dg.Attempts, &dg.CreatedAt, &dg.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	dg.Content = StringPtr(content)
	dg.SqlarName = StringPtr(sqlarName)
	dg.Error = StringPtr(digestError)

	return &dg, nil
}

// GetDigestsForFile retrieves all digests for a file
func (d *DB) GetDigestsForFile(filePath string) ([]Digest, error) {
	query := `
		SELECT id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at
		FROM digests
		WHERE file_path = ?
		ORDER BY digester
	`

	rows, err := d.conn.Query(query, filePath)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var digests []Digest
	for rows.Next() {
		var dg Digest
		var content, sqlarName, digestError sql.NullString

		err := rows.Scan(
			&dg.ID, &dg.FilePath, &dg.Digester, &dg.Status, &content,
			&sqlarName, &digestError, &dg.Attempts, &dg.CreatedAt, &dg.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		dg.Content = StringPtr(content)
		dg.SqlarName = StringPtr(sqlarName)
		dg.Error = StringPtr(digestError)

		digests = append(digests, dg)
	}

	return digests, nil
}

// GetDigestByFileAndDigester retrieves a specific digest
func (d *DB) GetDigestByFileAndDigester(filePath, digester string) (*Digest, error) {
	query := `
		SELECT id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at
		FROM digests
		WHERE file_path = ? AND digester = ?
	`

	row := d.conn.QueryRow(query, filePath, digester)

	var dg Digest
	var content, sqlarName, digestError sql.NullString

	err := row.Scan(
		&dg.ID, &dg.FilePath, &dg.Digester, &dg.Status, &content,
		&sqlarName, &digestError, &dg.Attempts, &dg.CreatedAt, &dg.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	dg.Content = StringPtr(content)
	dg.SqlarName = StringPtr(sqlarName)
	dg.Error = StringPtr(digestError)

	return &dg, nil
}

// CreateDigest creates a new digest record
func (d *DB) CreateDigest(ctx context.Context, dg *Digest) error {
	if dg.ID == "" {
		dg.ID = uuid.New().String()
	}
	now := NowMs()
	if dg.CreatedAt == 0 {
		dg.CreatedAt = now
	}
	if dg.UpdatedAt == 0 {
		dg.UpdatedAt = now
	}

	query := `
		INSERT INTO digests (id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`

	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(query,
			dg.ID, dg.FilePath, dg.Digester, dg.Status, dg.Content,
			dg.SqlarName, dg.Error, dg.Attempts, dg.CreatedAt, dg.UpdatedAt,
		)
		return err
	})
}

// UpdateDigest updates an existing digest
func (d *DB) UpdateDigest(ctx context.Context, dg *Digest) error {
	dg.UpdatedAt = NowMs()

	query := `
		UPDATE digests
		SET status = ?, content = ?, sqlar_name = ?, error = ?, attempts = ?, updated_at = ?
		WHERE id = ?
	`

	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(query,
			dg.Status, dg.Content, dg.SqlarName, dg.Error, dg.Attempts, dg.UpdatedAt, dg.ID,
		)
		return err
	})
}

// UpsertDigest creates or updates a digest
func (d *DB) UpsertDigest(ctx context.Context, dg *Digest) error {
	now := NowMs()
	if dg.ID == "" {
		dg.ID = GeneratePathHash(dg.FilePath) + "-" + dg.Digester
	}
	dg.UpdatedAt = now

	query := `
		INSERT INTO digests (id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		ON CONFLICT(file_path, digester) DO UPDATE SET
			status = excluded.status,
			content = excluded.content,
			sqlar_name = excluded.sqlar_name,
			error = excluded.error,
			attempts = excluded.attempts,
			updated_at = excluded.updated_at
	`

	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec(query,
			dg.ID, dg.FilePath, dg.Digester, dg.Status, dg.Content,
			dg.SqlarName, dg.Error, dg.Attempts, now, dg.UpdatedAt,
		)
		return err
	})
}

// DeleteDigest removes a digest
func (d *DB) DeleteDigest(ctx context.Context, id string) error {
	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec("DELETE FROM digests WHERE id = ?", id)
		return err
	})
}

// DeleteDigestsForFile removes all digests for a file
func (d *DB) DeleteDigestsForFile(ctx context.Context, filePath string) error {
	return d.Write(ctx, func(tx *sql.Tx) error {
		_, err := tx.Exec("DELETE FROM digests WHERE file_path = ?", filePath)
		return err
	})
}

// ResetDigesterAll resets all digests of a specific type
func (d *DB) ResetDigesterAll(ctx context.Context, digester string) (int64, error) {
	var affected int64
	err := d.Write(ctx, func(tx *sql.Tx) error {
		result, err := tx.Exec(`
			UPDATE digests
			SET status = 'todo', error = NULL, attempts = 0, updated_at = ?
			WHERE digester = ?
		`, NowMs(), digester)
		if err != nil {
			return err
		}
		affected, err = result.RowsAffected()
		return err
	})
	if err != nil {
		return 0, err
	}
	return affected, nil
}

// GetDigestStats returns statistics about digests
func (d *DB) GetDigestStats() (map[string]interface{}, error) {
	stats := make(map[string]interface{})

	// Count by status
	rows, err := d.conn.Query(`
		SELECT status, COUNT(*) as count
		FROM digests
		GROUP BY status
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	byStatus := make(map[string]int64)
	for rows.Next() {
		var status string
		var count int64
		if err := rows.Scan(&status, &count); err != nil {
			return nil, err
		}
		byStatus[status] = count
	}
	stats["byStatus"] = byStatus

	// Count by digester
	rows, err = d.conn.Query(`
		SELECT digester, status, COUNT(*) as count
		FROM digests
		GROUP BY digester, status
		ORDER BY digester
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	byDigester := make(map[string]map[string]int64)
	for rows.Next() {
		var digester, status string
		var count int64
		if err := rows.Scan(&digester, &status, &count); err != nil {
			return nil, err
		}
		if _, ok := byDigester[digester]; !ok {
			byDigester[digester] = make(map[string]int64)
		}
		byDigester[digester][status] = count
	}
	stats["byDigester"] = byDigester

	// Total
	var total int64
	err = d.conn.QueryRow("SELECT COUNT(*) FROM digests").Scan(&total)
	if err != nil {
		return nil, err
	}
	stats["total"] = total

	return stats, nil
}

// GetPendingDigests retrieves digests that need processing
func (d *DB) GetPendingDigests(limit int) ([]Digest, error) {
	query := `
		SELECT id, file_path, digester, status, content, sqlar_name, error, attempts, created_at, updated_at
		FROM digests
		WHERE status IN ('todo', 'failed')
		  AND attempts < 3
		ORDER BY created_at ASC
		LIMIT ?
	`

	rows, err := d.conn.Query(query, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var digests []Digest
	for rows.Next() {
		var dg Digest
		var content, sqlarName, digestError sql.NullString

		err := rows.Scan(
			&dg.ID, &dg.FilePath, &dg.Digester, &dg.Status, &content,
			&sqlarName, &digestError, &dg.Attempts, &dg.CreatedAt, &dg.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}

		dg.Content = StringPtr(content)
		dg.SqlarName = StringPtr(sqlarName)
		dg.Error = StringPtr(digestError)

		digests = append(digests, dg)
	}

	return digests, nil
}

// GetDistinctDigesters returns all unique digester names
func (d *DB) GetDistinctDigesters() ([]string, error) {
	rows, err := d.conn.Query("SELECT DISTINCT digester FROM digests ORDER BY digester")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var digesters []string
	for rows.Next() {
		var digester string
		if err := rows.Scan(&digester); err != nil {
			return nil, err
		}
		digesters = append(digesters, digester)
	}

	return digesters, nil
}

// ListDigestsForPath returns all digests for a file path (alias for GetDigestsForFile)
func (d *DB) ListDigestsForPath(filePath string) []Digest {
	digests, err := d.GetDigestsForFile(filePath)
	if err != nil {
		return nil
	}
	return digests
}

// GetDigestByPathAndDigester returns a digest by file path and digester name
func (d *DB) GetDigestByPathAndDigester(filePath, digester string) *Digest {
	dg, err := d.GetDigestByFileAndDigester(filePath, digester)
	if err != nil {
		return nil
	}
	return dg
}

// UpdateDigestMap updates a digest using a map of fields
func (d *DB) UpdateDigestMap(ctx context.Context, id string, fields map[string]interface{}) error {
	dg, err := d.GetDigestByID(id)
	if err != nil || dg == nil {
		return err
	}

	// Apply updates
	if status, ok := fields["status"].(string); ok {
		dg.Status = status
	}
	if content, ok := fields["content"].(string); ok {
		dg.Content = &content
	}
	if sqlarName, ok := fields["sqlar_name"].(string); ok {
		dg.SqlarName = &sqlarName
	}
	if errorStr, ok := fields["error"]; ok {
		if errorStr == nil {
			dg.Error = nil
		} else if s, ok := errorStr.(string); ok {
			dg.Error = &s
		} else if sp, ok := errorStr.(*string); ok {
			dg.Error = sp
		}
	}
	if attempts, ok := fields["attempts"].(int); ok {
		dg.Attempts = attempts
	}
	if updatedAt, ok := fields["updated_at"].(int64); ok {
		dg.UpdatedAt = updatedAt
	}

	return d.UpdateDigest(ctx, dg)
}

// GetFilesWithPendingDigests returns file paths that have pending or failed digests
// Only returns inbox files - library files don't get auto-digested
func (d *DB) GetFilesWithPendingDigests() []string {
	query := `
		SELECT DISTINCT d.file_path
		FROM digests d
		INNER JOIN files f ON f.path = d.file_path
		WHERE d.status IN ('todo', 'failed')
		  AND d.attempts < 3
		  AND (d.file_path = 'inbox' OR d.file_path LIKE 'inbox/%')
		ORDER BY d.created_at ASC
		LIMIT 100
	`

	rows, err := d.conn.Query(query)
	if err != nil {
		return nil
	}
	defer rows.Close()

	var paths []string
	for rows.Next() {
		var path string
		if err := rows.Scan(&path); err != nil {
			continue
		}
		paths = append(paths, path)
	}

	return paths
}
